packages/salesforce-adapter/src/filters/foreign_key_references.ts

```diff
@@ -0,0 +1,126 @@
+* the License.  You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+import {
+  ElemID, Element, Field, isObjectType, InstanceElement, ObjectType, ReferenceExpression,
+  isInstanceElement, Value, isReferenceExpression,
+} from '@salto-io/adapter-api'
+import _ from 'lodash'
+import { transformValues, TransformFunc } from '@salto-io/adapter-utils'
+import { values, collections } from '@salto-io/lowerdash'
+import { FilterCreator } from '../filter'
+import { groupByAPIName, ApiNameMapping } from './instance_references'
+import { FOREIGN_KEY_DOMAIN } from '../constants'
+import { apiName } from '../transformers/transformer'
+
+const { makeArray } = collections.array
+
+/**
+ * Convert foreign-key-domain annotations to reference expressions using the known metadata types.
+ *
+ * @param elements      The fetched elements
+ * @param typeToElemID  Known element ids by metadata type
+ */
+const convertAnnotationsToReferences = (
+  elements: Element[],
+  typeToElemID: Record<string, ElemID>,
+): void => {
+  const resolveTypeReference = (ref: string | ReferenceExpression):
+    string | ReferenceExpression => {
+    if (_.isString(ref)) {
+      const referenceElemId = typeToElemID[ref]
+      if (referenceElemId !== undefined) {
+        return new ReferenceExpression(referenceElemId)
+      }
+    }
+    return ref
+  }
+
+  elements
+    .filter(isObjectType)
+    .flatMap((obj: ObjectType) => Object.values(obj.fields))
+    .filter((field: Field) => field.annotations[FOREIGN_KEY_DOMAIN] !== undefined)
+    .forEach((field: Field): void => {
+      field.annotations[FOREIGN_KEY_DOMAIN] = field.annotations[FOREIGN_KEY_DOMAIN]
+        .map(resolveTypeReference)
+    })
+}
+
+/**
+ * Resolve references using the mapping genrated from the Salesforce DescribeValueType API.
+ *
+ * @param instance                The current instance being modified
+ * @param apiNameToElemIDs        Known element ids, mapped by API name and metadata type
+ */
+const resolveReferences = (
+  instance: InstanceElement,
+  apiNameToElemIDs: ApiNameMapping,
+): void => {
+  const shouldReplace = (field: Field): boolean => (
+    makeArray(field.annotations[FOREIGN_KEY_DOMAIN]).some(isReferenceExpression)
+  )
+
+  const replacePrimitive = (val: Value, field: Field): Value => {
+    if (val === undefined || !_.isString(val)) {
+      return val
+    }
+
+    const refTarget = field.annotations[FOREIGN_KEY_DOMAIN]
+      .filter(isReferenceExpression)
+      .map((ref: ReferenceExpression) => apiNameToElemIDs[val]?.[
+        ref.elemId.typeName
+      ])
+      .find(values.isDefined)
```

Comment:
again, let typescript infer the types
```suggestion
    const refTarget = field.annotations[FOREIGN_KEY_DOMAIN]
      .filter(isReferenceExpression)
      .map(ref => apiNameToElemIDs[val]?.[ref.elemId.typeName])
      .find(values.isDefined)
```

Reply:
this one seems needed